lib/fetch_blocks_list.js

```diff
@@ -0,0 +1,239 @@
+      key = responseKey;
+      isTruncated = responseIsTruncated;
+    }
+    return data
+      .map(x => x.Key)
+      .filter(x => x !== undefined);
+  }
+
+  /**
+   * @param {Array} keys Array of S3 objects' keys
+   * @param {string} blockchain Blockchain name to filter the keys by
+   * @returns {Array} Array of S3 objects that match the key
+  */
+  filterS3Keys(keys, blockchain) {
+    return keys.filter((key) => key && key.startsWith(blockchain) && key.endsWith('.csv'));
+  }
+
+  /**
+   * @param {object} response Object from S3
+   * @returns {Array} Array of data from S3 object
+   */
+  async processDataIntoArray(response, key) {
+    const responseStream = response.Body;
+    const csvData = [];
+    return new Promise((resolve, reject) => {
+      if (responseStream.statusCode !== 200) {
+        logger.error(`Error processing data for ${key}: ${responseStream.status_code}`);
+        reject(new Error());
+      }
+      responseStream
+        .pipe(csv())
+        .on('data', (data) => {
+          const blockNumber = Object.values(data)[0];
+          csvData.push(blockNumber);
+        })
+        .on('end', () => {
+          logger.info(`Processed data for ${key}`);
+          resolve(csvData);
+        });
+    });
+  }
+
+  /**
+   * @param {Array} keys S3 object keys
+   * @returns {Array} Array of block numbers in ascending order, from all S3 object keys
+   */
+  async processAllS3Data(keys) {
+    const result = {};
+    for (const key of keys) {
+      const currObject = await this.s3Client.send(new GetObjectCommand({
+        Bucket: S3_BUCKET,
+        Key: key
+      }));
+      const csvData = await this.processDataIntoArray(currObject, key);
+      result[key] = csvData
+        .sort((a, b) => a - b)
+        .map(x => parseInt(x));
+    }
+    return result;
+  }
+
+  /**
+   * @param {number} start Start block number
+   * @param {number} end End block number
+   * @param {number} interval Maximum distance between blocks to be fetched
+   * @returns {Array} Array of intervals
+   */
+  breakdownInterval(start, end) {
+    const result = [];
+    if (end - start <= EXPORT_BLOCKS_LIST_MAX_INTERVAL) {
+      result.push([start, end]);
+      return result;
+    }
+    let currStart = start;
+    let currEnd = start + EXPORT_BLOCKS_LIST_MAX_INTERVAL - 1;
+    while (currEnd < end) {
+      result.push([currStart, currEnd]);
+      currStart = currEnd + 1;
+      currEnd = currStart + EXPORT_BLOCKS_LIST_MAX_INTERVAL - 1;
+    }
+    result.push([currStart, end]);
+    return result;
+  }
+
+  //TODO: How to track via ZooKeeper
+  /**
+   * @param {Array} data Sorted array of block numbers in ascending order
+   * @returns {Array} Array of intervals
+   */
+  makeIntervals(data) {
+    if (data.length === 0) {
+      return [];
+    }
+    const intervals = [];
+    let start = data[0];
+    let end = data[0];
+    for (let i = 1; i < data.length; i++) {
+      if (data[i] - data[i - 1] === 1) {
```

Comment:
can we assert that `data[i] > data[i-1]`?

Reply:
Yes, the array should be sorted in ascending order beforehand
```js
---processAllS3Data---
result[key] = csvData
  .sort((a, b) => a - b)
  .map(x => parseInt(x));
```